### DISCUSSION TITLE
Fluxcd not able to deploy workload cluster using manifest stored in github

### DISCUSSION THREAD

**Discussion by abashir236** [fluxcd/flux2#3047] [Q&A] [answered]
I am going to post the directory tree below but I am using KinD which has a management cluster and attempting to deploy a workload cluster. I have created the source and kustomization files however the error I am getting suggests that the clusterClass is not being created and thus the cluster cannot be as well. The error is: 

`capi-quickstart		False	False	Cluster/default/capi-quickstart dry-run failed, reason: InternalError, error: admission webhook "default.cluster.cluster.x-k8s.io" denied the request: Internal error occurred: Cluster capi-quickstart can't be validated. ClusterClass quick-start can not be retrieved: ClusterClass.cluster.x-k8s.io "quick-start" not found`

<img width="454" alt="Screen Shot 2022-08-28 at 10 14 18 PM" src="https://user-images.githubusercontent.com/41330628/187109671-7ab95354-5ba4-40f8-ac79-cedc25320501.png">

I was able to deploy the cluster manually using kubectl and confirmed that flux is now monitoring the repo and I have confirmed this by reducing and increasing num of workers and pushing manifest to github and flux reconciles and scales up or down accordingly so all good there. Just wondering why it would not be able to digest the entire manifest file and deploy it for me.

The manifest was generated by running: ```
clusterctl generate cluster capi-quickstart --flavor development \
  --kubernetes-version v1.24.0 \
  --control-plane-machine-count=3 \
  --worker-machine-count=3 \
  > capi-quickstart.yaml
```

Thanks in advance.

**Comment by kingdonb** [comment, marked as answer]
Greetings, the ClusterClass has to be deployed in order for it to be usable:

> `ClusterClass quick-start can not be retrieved: ClusterClass.cluster.x-k8s.io "quick-start" not found`

This error indicates the Admission Controller for CAPI rejected the cluster because the clusterclass was not defined yet.

It worked with kubectl I'm assuming because of random chance, the order that kubectl applied the resources happened to be correct, and the behavior is asynchronous. So this was a lucky roll of the dice. Because of Flux's validating behavior, validations are strict and Flux knows so much about which resources can be applied in which order to produce a valid outcome – it can figure out for you that Namespaces must go before Namespaced resources, and that ConfigMaps go before Deployments, etc. for common pairs that must be performed in order. Flux knows many of these relationships, but it cannot get every one right. I do think that an improvement is possible so you do not have to declare ClusterClass in a separate Kustomization to guarantee the order. But for now, this is not implemented, so since ClusterClass is used in the Cluster, it should be declared in a separate Kustomization that gets applied one before the other.

You can set up Flux to apply resources in order with the `dependsOn` feature of Kustomize controller. This way you can be sure the ClusterClass is defined and accepted by the cluster before it is used, and you should be able to avoid this failure that way.

Please see this doc for more: https://fluxcd.io/flux/components/kustomize/kustomization/#kustomization-dependencies

**Reply by kingdonb** [reply]
I thought this would be a common enough use case (the Cluster API project is well known and ClusterClass+Cluster are a stable dependency pair which always has this order) that I opened an issue to see whether Flux maintainers think this can be solved without making users tease this issue out in their Flux manifest repository structure:

* https://github.com/fluxcd/kustomize-controller/issues/719

Hopefully this makes sense. I think we want ClusterAPI users to have a good experience, so we might just add this logic. Let's see what others think.

**Reply by abashir236** [reply]
FYI the dependsOn feature is exactly what I ended up using eventually but glad to see that this is something that will be considered. 

Thanks.

**Reply by kingdonb** [reply]
@abashir236 It has been considered and accepted, this suffering will not be necessary for future generations of Flux users:

* https://github.com/fluxcd/pkg/pull/347

Thank you for the report!
